File: src/com/zl/java/easyfactory/EasyFactory.java
package com.zl.java.easyfactory;

/**
 * Created by rudy
 * 2016
 * 16-12-6
 * 下午11:18
 */
public class EasyFactory {


  public static Api factory(int n) {
    Api api = null;
    switch (n) {
      case 1:
        api = new ImplA();
        break;
      case 2:
        api = new ImplB();
        break;
      default:
        throw new RuntimeException("error n");
    }
    return api;
  }
}

File: src/com/zl/kotlin/easyfactory/easyfactory.kt
package com.zl.kotlin.easyfactory

/**
 * Created by rudy
 * 2016
 * 16-12-6
 * 下午11:26
 */

interface Api {
    fun operation(s: String)
}

class Impla : Api {
    override fun operation(s: String) {
        println("this is Impla " + s)
    }
}

class Implb : Api {
    override fun operation(s: String) {
        println("this is Implb " + s)
    }
}

fun factory(n: Int): Api {
    var api: Api
    when (n) {
        1 -> api = Impla()
        2 -> api = Implb()
        else -> throw IllegalArgumentException("errror arg")
    }
    return api
}
